File: src/main/java/com/akagiyui/drive/service/impl/FolderServiceImpl.java
package com.akagiyui.drive.service.impl;

import com.akagiyui.common.ResponseEnum;
import com.akagiyui.common.exception.CustomException;
import com.akagiyui.drive.entity.Folder;
import com.akagiyui.drive.entity.User;
import com.akagiyui.drive.model.response.FolderResponse;
import com.akagiyui.drive.repository.FolderRepository;
import com.akagiyui.drive.service.FolderService;
import com.akagiyui.drive.service.UserService;
import jakarta.annotation.Resource;
import org.jetbrains.annotations.NotNull;
import org.springframework.stereotype.Service;
import org.springframework.util.StringUtils;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

/**
 * 文件夹 服务实现类
 *
 * @author AkagiYui
 */
@Service
public class FolderServiceImpl implements FolderService {

    @Resource
    private FolderRepository folderRepository;

    @Resource
    private UserService userService;

    @Override
    public @NotNull Folder createFolder(@NotNull String name, String parentId) {
        User user = userService.getUser();

        parentId = StringUtils.hasText(parentId) ? parentId : null;
        Folder parentFolder = parentId != null ? folderRepository.findById(parentId).orElseThrow(() -> new CustomException(ResponseEnum.NOT_FOUND)) : null;

        if (folderRepository.existsByNameAndUserIdAndParentId(name, user.getId(), parentId)) {
            throw new CustomException(ResponseEnum.FOLDER_EXIST);
        }

        Folder folder = new Folder()
            .setName(name)
            .setParent(parentFolder)
            .setUser(user);
        folderRepository.save(folder);
        return folder;
    }

    @Override
    public @NotNull Folder createFolder(@NotNull String name) {
        return createFolder(name, null);
    }

    @Override
    public @NotNull List<Folder> getSubFolders(String parentId) {
        User user = userService.getUser();
        parentId = StringUtils.hasText(parentId) ? parentId : null;

        return folderRepository.findByUserIdAndParentId(user.getId(), parentId);
    }

    @Override
    public @NotNull List<FolderResponse> getFolderChain(@NotNull String folderId) {
        User user = userService.getUser();
        if (!StringUtils.hasText(folderId)) {
            return new ArrayList<>();
        }

        Folder folder = folderRepository.findById(folderId).orElseThrow(() -> new CustomException(ResponseEnum.NOT_FOUND));
        if (!Objects.equals(folder.getUser().getId(), user.getId())) {
            throw new CustomException(ResponseEnum.NOT_FOUND);
        }

        List<FolderResponse> folderChain = new ArrayList<>();
        while (folder != null) {
            folderChain.addFirst(new FolderResponse(folder));
            folder = folder.getParent();
        }

        return folderChain;
    }
}

File: app/src/main/kotlin/com/akagiyui/drive/service/impl/FolderServiceImpl.kt
package com.akagiyui.drive.service.impl

import com.akagiyui.common.ResponseEnum
import com.akagiyui.common.delegate.LoggerDelegate
import com.akagiyui.common.exception.BusinessException
import com.akagiyui.common.utils.hasText
import com.akagiyui.drive.entity.Folder
import com.akagiyui.drive.entity.User
import com.akagiyui.drive.model.response.folder.FolderResponse
import com.akagiyui.drive.repository.FolderRepository
import com.akagiyui.drive.service.FolderService
import com.akagiyui.drive.service.UserFileService
import jakarta.annotation.Resource
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.context.annotation.Lazy
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

/**
 * 文件夹服务实现类
 *
 * @author AkagiYui
 */
@Service
class FolderServiceImpl @Autowired constructor(
    private val folderRepository: FolderRepository,
) : FolderService {
    private val log by LoggerDelegate()

    @Lazy
    @Resource
    private lateinit var userFileService: UserFileService

    override fun createFolder(user: User, name: String, parentId: String?): Folder {
        val resolvedParentId = parentId?.takeIf { it.hasText() }
        val parentFolder = resolvedParentId?.let {
            folderRepository.findById(it).orElseThrow { BusinessException(ResponseEnum.NOT_FOUND) }
        }

        if (folderRepository.existsByNameAndUserIdAndParentId(name, user.id, resolvedParentId)) {
            throw BusinessException(ResponseEnum.FOLDER_EXIST)
        }

        val folder = Folder().apply {
            this.name = name
            this.user = user
            this.parent = parentFolder
        }
        folderRepository.save(folder)
        return folder
    }

    override fun getSubFolders(userId: String, parentId: String?): List<Folder> {
        val resolvedParentId = parentId?.takeIf { it.hasText() }
        return folderRepository.findByUserIdAndParentId(userId, resolvedParentId)
    }

    override fun getFolderChain(userId: String, folderId: String): List<FolderResponse> {
        if (!folderId.hasText()) {
            return listOf()
        }

        val folder = folderRepository.findById(folderId).orElseThrow { BusinessException(ResponseEnum.NOT_FOUND) }
        if (folder.user.id != userId) {
            throw BusinessException(ResponseEnum.NOT_FOUND)
        }

        val folderChain = mutableListOf<FolderResponse>()
        var currentFolder: Folder? = folder
        while (currentFolder != null) {
            folderChain.add(0, FolderResponse(currentFolder))
            currentFolder = currentFolder.parent
        }

        return folderChain.toList()
    }

    override fun getFolderById(folderId: String): Folder {
        return folderRepository.findById(folderId).orElseThrow { BusinessException(ResponseEnum.NOT_FOUND) }
    }

    @Transactional
    override fun deleteFolder(userId: String, folderId: String) {
        val folder = folderRepository.findById(folderId).orElseThrow { BusinessException(ResponseEnum.NOT_FOUND) }
        if (folder.user.id != userId) {
            throw BusinessException(ResponseEnum.NOT_FOUND)
        }

        // 删除文件夹下的文件和子文件夹
        folder.files.forEach {
            userFileService.userDeleteFile(userId, it.id)
        }
        folder.subFolders.forEach {
            deleteFolder(userId, it.id)
        }

        folderRepository.delete(folder)
        log.debug("delete folder: $folderId")
    }

    override fun rename(userId: String, folderId: String, newName: String) {
        val folder = folderRepository.findByUserIdAndId(userId, folderId)
            ?: throw BusinessException(ResponseEnum.NOT_FOUND)
        folder.name = newName
        folderRepository.save(folder)
    }

    override fun moveFolder(userId: String, folderId: String, parentId: String?) {
        val folder = folderRepository.findById(folderId).orElseThrow { BusinessException(ResponseEnum.NOT_FOUND) }
        if (folder.user.id != userId) {
            throw BusinessException(ResponseEnum.NOT_FOUND)
        }

        check(parentId != folderId) { "not allowed to move folder to itself" }
        val parentFolder = parentId.hasText {
            folderRepository.findById(it).orElseThrow { BusinessException(ResponseEnum.NOT_FOUND) }
        }

        folder.parent = parentFolder
        folderRepository.save(folder)
    }
}
